src/editable_tree/mod.rs

```diff
@@ -584,45 +585,1164 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
+        let root = start_tree.add_to_arena(&arena);
+        let mut editable_tree = DAG::new(&arena, root, start_cursor_location);
+
+        assert_eq!(
+            expected_edit_result,
+            editable_tree.execute_action(action),
+            "Not equal in action result"
+        );
+        assert_eq!(expected_tree, editable_tree.root(), "Not equal in tree.");
+        assert_eq!(
+            expected_cursor_location, editable_tree.current_cursor_path,
+            "Not equal in cursor location."
+        );
+    }
+
+    #[test]
+    fn dag_root_insertchild() {
+        let start_tree = TestJSON::Array(vec![]);
+        let start_cursor_location = CursorPath::from_vec(vec![]);
+        let action = Action::InsertChild('f');
+
+        let expected_edit_result: (bool, Result<EditSuccess, EditErr>) = (
+            false,
+            Ok(EditSuccess::InsertChild {
+                c: 'f',
+                name: "false".to_string(),
+            }),
+        );
+        let expected_tree = TestJSON::Array(vec![TestJSON::False]);
+        let expected_cursor_location = CursorPath::from_vec(vec![]);
+
+        run_test(
+            start_tree,
+            start_cursor_location,
+            action,
+            expected_edit_result,
+            expected_tree,
+            expected_cursor_location,
+        );
+
+        //DAG level == 1
+        let start_tree = TestJSON::Array(vec![TestJSON::Array(vec![]), TestJSON::True]);
+        let start_cursor_location = CursorPath::from_vec(vec![]);
+        let action = Action::InsertChild('n');
+
+        let expected_edit_result: (bool, Result<EditSuccess, EditErr>) = (
+            false,
+            Ok(EditSuccess::InsertChild {
+                c: 'n',
+                name: "null".to_string(),
+            }),
+        );
+        let expected_tree = TestJSON::Array(vec![
+            TestJSON::Array(vec![]),
+            TestJSON::True,
+            TestJSON::Null,
+        ]);
+        let expected_cursor_location = CursorPath::from_vec(vec![]);
+
+        run_test(
+            start_tree,
+            start_cursor_location,
+            action,
+            expected_edit_result,
+            expected_tree,
+            expected_cursor_location,
+        );
+    }
+
+    #[test]
+    fn dag_root_insertbefore() {
+        let start_tree = TestJSON::Array(vec![]);
+        let start_cursor_location = CursorPath::from_vec(vec![]);
+        let action = Action::InsertBefore('f');
+
+        let expected_edit_result: (bool, Result<EditSuccess, EditErr>) =
+            (false, Err(EditErr::AddSiblingToRoot));
+        let expected_tree = TestJSON::Array(vec![]);
+        let expected_cursor_location = CursorPath::from_vec(vec![]);
+
+        run_test(
+            start_tree,
+            start_cursor_location,
+            action,
+            expected_edit_result,
+            expected_tree,
+            expected_cursor_location,
+        );
```

Comment:
For actions that fail, we would always expect `expected_tree` to equal `start_tree` and `expected_cursor_location` to equal `start_cursor_location`, as well as `expected_edit_result` to have the form `(false, Err(<some_err>))`.  So we may as well make another function that runs a test that we expect to fail.  So this test would become:
```rust
#[test]
mod tests {
    fn run_test(...) {
        // --snip--
    }

    /// Run a test that is expected to fail
    fn run_test_err(
        tree: TestJSON,
        cursor_location: CursorPath,
        action: Action,
        exp_error: EditErr,
    ) {
        let arena: Arena<JSON> = Arena::new();
        let root = tree.add_to_arena(&arena);
        let mut editable_tree = DAG::new(&arena, root, cursor_location);

        assert_eq!(
            (false, Err(exp_error)),
            editable_tree.execute_action(action),
            "Unexpected result of the action."
        );
        assert_eq!(
            tree,
            editable_tree.root(),
            "Failed action changed the tree."
        );
        assert_eq!(
            editable_tree.current_cursor_path, cursor_location,
            "Failed action changed the cursor location."
        );
    }

    // --snip--

    #[cfg(test)]
    fn dag_root_insertbefore() {
        run_test_err(
            TestJSON::Array(vec![]),
            CursorPath::from_vec(vec![]),
            Action::InsertBefore('f'),
            EditErr::AddSiblingToRoot,
        );
    }
}
```